File: python/trainer.py
"""
Training pipeline for fine-tuning LLMs on financial data.

This module provides a flexible training framework supporting:
- LoRA and QLoRA fine-tuning
- Early stopping and checkpointing
- Learning rate scheduling
- Gradient accumulation
"""

import os
import json
import time
from datetime import datetime
from typing import Optional, Dict, Any, List, Callable, Tuple
from dataclasses import dataclass, field, asdict
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, Dataset
from torch.optim import AdamW
from torch.optim.lr_scheduler import CosineAnnealingLR, LinearLR, SequentialLR
import pandas as pd
import numpy as np

from .model import FinancialSentimentLoRA, LoRAConfig, count_parameters


@dataclass
class TrainingConfig:
    """Configuration for training."""
    # Model
    model_name: str = "bert-base-uncased"
    method: str = "lora"  # lora, qlora, prefix, full
    lora_rank: int = 8
    lora_alpha: float = 16.0
    lora_dropout: float = 0.1

    # Training
    epochs: int = 3
    batch_size: int = 16
    learning_rate: float = 2e-4
    weight_decay: float = 0.01
    warmup_ratio: float = 0.1
    max_grad_norm: float = 1.0

    # Early stopping
    early_stopping: bool = True
    patience: int = 3
    min_delta: float = 0.001

    # Checkpointing
    save_dir: str = "outputs"
    save_best_only: bool = True
    save_total_limit: int = 3

    # Logging
    log_steps: int = 10
    eval_steps: int = 100

    # Hardware
    device: str = "auto"
    mixed_precision: bool = True
    gradient_accumulation_steps: int = 1

    def __post_init__(self):
        if self.device == "auto":
            self.device = "cuda" if torch.cuda.is_available() else "cpu"


@dataclass
class TrainingMetrics:
    """Metrics collected during training."""
    epoch: int = 0
    step: int = 0
    train_loss: float = 0.0
    val_loss: float = 0.0
    val_accuracy: float = 0.0
    val_f1: float = 0.0
    learning_rate: float = 0.0
    best_val_loss: float = float("inf")
    best_epoch: int = 0
    history: List[Dict[str, float]] = field(default_factory=list)


class FinancialTextDataset(Dataset):
    """
    Dataset for financial text classification.

    Args:
        texts: List of text strings
        labels: List of integer labels
        tokenizer: Tokenizer function or object
        max_length: Maximum sequence length
    """

    def __init__(
        self,
        texts: List[str],
        labels: List[int],
        tokenizer: Optional[Callable] = None,
        max_length: int = 512
    ):
        self.texts = texts
        self.labels = labels
        self.max_length = max_length

        # Simple tokenizer if none provided
        self.tokenizer = tokenizer or self._simple_tokenizer

    def _simple_tokenizer(self, text: str) -> Dict[str, torch.Tensor]:
        """Simple character-level tokenizer for demonstration."""
        # In practice, use HuggingFace tokenizer
        tokens = [ord(c) % 10000 for c in text[:self.max_length]]
        tokens = tokens + [0] * (self.max_length - len(tokens))

        return {
            "input_ids": torch.tensor(tokens),
            "attention_mask": torch.tensor([1] * len(text[:self.max_length]) +
                                          [0] * (self.max_length - len(text[:self.max_length])))
        }

    def __len__(self) -> int:
        return len(self.texts)

    def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
        text = self.texts[idx]
        label = self.labels[idx]

        encoding = self.tokenizer(text)
        encoding["labels"] = torch.tensor(label)

        return encoding


class EarlyStopping:
    """
    Early stopping to terminate training when validation loss stops improving.

    Args:
        patience: Number of epochs to wait before stopping
        min_delta: Minimum change to qualify as improvement
        mode: "min" for loss, "max" for accuracy
    """

    def __init__(
        self,
        patience: int = 3,
        min_delta: float = 0.001,
        mode: str = "min"
    ):
        self.patience = patience
        self.min_delta = min_delta
        self.mode = mode
        self.counter = 0
        self.best_score = None
        self.early_stop = False

    def __call__(self, score: float) -> bool:
        """
        Check if training should stop.

        Args:
            score: Current metric value

        Returns:
            True if training should stop
        """
        if self.best_score is None:
            self.best_score = score
            return False

        if self.mode == "min":
            improved = score < self.best_score - self.min_delta
        else:
            improved = score > self.best_score + self.min_delta

        if improved:
            self.best_score = score
            self.counter = 0
        else:
            self.counter += 1
            if self.counter >= self.patience:
                self.early_stop = True

        return self.early_stop


class FineTuningTrainer:
    """
    Trainer for fine-tuning LLMs on financial tasks.

    Supports LoRA, QLoRA, and full fine-tuning with various
    optimization techniques.

    Example:
        >>> trainer = FineTuningTrainer(model_name="finbert", method="lora")
        >>> trainer.train(train_data, val_data, epochs=3)
        >>> metrics = trainer.evaluate(test_data)
    """

    def __init__(
        self,
        model_name: str = "bert-base-uncased",
        method: str = "lora",
        lora_rank: int = 8,
        lora_alpha: float = 16.0,
        learning_rate: float = 2e-4,
        config: Optional[TrainingConfig] = None
    ):
        """
        Initialize trainer.

        Args:
            model_name: Name of base model
            method: Fine-tuning method (lora, qlora, prefix, full)
            lora_rank: LoRA rank
            lora_alpha: LoRA scaling factor
            learning_rate: Initial learning rate
            config: Full training configuration
        """
        if config is None:
            config = TrainingConfig(
                model_name=model_name,
                method=method,
                lora_rank=lora_rank,
                lora_alpha=lora_alpha,
                learning_rate=learning_rate
            )

        self.config = config
        self.device = torch.device(config.device)

        # Initialize model
        self.model = self._create_model()
        self.model.to(self.device)

        # Metrics tracking
        self.metrics = TrainingMetrics()

        # Early stopping
        self.early_stopping = EarlyStopping(
            patience=config.patience,
            min_delta=config.min_delta
        ) if config.early_stopping else None

        # Mixed precision
        self.scaler = torch.cuda.amp.GradScaler() if (
            config.mixed_precision and config.device == "cuda"
        ) else None

    def _create_model(self) -> nn.Module:
        """Create model based on configuration."""
        # For this implementation, we use our custom LoRA model
        # In practice, you'd load a HuggingFace model and apply PEFT
        model = FinancialSentimentLoRA(
            base_dim=768,
            lora_rank=self.config.lora_rank,
            num_classes=3
        )

        # Log parameter counts
        params = count_parameters(model)
        print(f"Model created:")
        print(f"  Trainable parameters: {params['trainable']:,}")
        print(f"  Total parameters: {params['total']:,}")
        print(f"  Trainable %: {params['trainable_percent']:.2f}%")

        return model

    def _create_optimizer(self) -> torch.optim.Optimizer:
        """Create optimizer with parameter groups."""
        # Separate LoRA and non-LoRA parameters
        lora_params = []
        other_params = []

        for name, param in self.model.named_parameters():
            if not param.requires_grad:
                continue
            if "lora" in name.lower():
                lora_params.append(param)
            else:
                other_params.append(param)

        param_groups = [
            {"params": lora_params, "lr": self.config.learning_rate},
            {"params": other_params, "lr": self.config.learning_rate * 0.1}
        ]

        return AdamW(
            param_groups,
            weight_decay=self.config.weight_decay
        )

    def _create_scheduler(
        self,
        optimizer: torch.optim.Optimizer,
        num_training_steps: int
    ):
        """Create learning rate scheduler with warmup."""
        num_warmup_steps = int(num_training_steps * self.config.warmup_ratio)

        warmup_scheduler = LinearLR(
            optimizer,
            start_factor=0.1,
            end_factor=1.0,
            total_iters=num_warmup_steps
        )

        decay_scheduler = CosineAnnealingLR(
            optimizer,
            T_max=num_training_steps - num_warmup_steps
        )

        return SequentialLR(
            optimizer,
            schedulers=[warmup_scheduler, decay_scheduler],
            milestones=[num_warmup_steps]
        )

    def train(
        self,
        train_data: pd.DataFrame,
        val_data: pd.DataFrame,
        epochs: Optional[int] = None,
        callbacks: Optional[List[Callable]] = None
    ) -> TrainingMetrics:
        """
        Train the model.

        Args:
            train_data: Training DataFrame with 'text' and 'label' columns
            val_data: Validation DataFrame
            epochs: Number of epochs (overrides config)
            callbacks: Optional list of callback functions

        Returns:
            TrainingMetrics with training history
        """
        epochs = epochs or self.config.epochs

        # Create datasets
        train_dataset = FinancialTextDataset(
            train_data["text"].tolist(),
            train_data["label"].tolist()
        )
        val_dataset = FinancialTextDataset(
            val_data["text"].tolist(),
            val_data["label"].tolist()
        )

        # Create data loaders
        train_loader = DataLoader(
            train_dataset,
            batch_size=self.config.batch_size,
            shuffle=True
        )
        val_loader = DataLoader(
            val_dataset,
            batch_size=self.config.batch_size
        )

        # Setup training
        optimizer = self._create_optimizer()
        num_training_steps = len(train_loader) * epochs
        scheduler = self._create_scheduler(optimizer, num_training_steps)

        # Training loop
        print(f"\nStarting training for {epochs} epochs...")
        start_time = time.time()

        for epoch in range(epochs):
            self.metrics.epoch = epoch + 1

            # Train epoch
            train_loss = self._train_epoch(train_loader, optimizer, scheduler)

            # Validate
            val_metrics = self._validate(val_loader)

            # Update metrics
            self.metrics.train_loss = train_loss
            self.metrics.val_loss = val_metrics["loss"]
            self.metrics.val_accuracy = val_metrics["accuracy"]
            self.metrics.learning_rate = scheduler.get_last_lr()[0]

            self.metrics.history.append({
                "epoch": epoch + 1,
                "train_loss": train_loss,
                "val_loss": val_metrics["loss"],
                "val_accuracy": val_metrics["accuracy"],
                "learning_rate": self.metrics.learning_rate
            })

            # Print progress
            print(f"Epoch {epoch + 1}/{epochs}: "
                  f"train_loss={train_loss:.4f}, "
                  f"val_loss={val_metrics['loss']:.4f}, "
                  f"val_acc={val_metrics['accuracy']:.4f}")

            # Check for best model
            if val_metrics["loss"] < self.metrics.best_val_loss:
                self.metrics.best_val_loss = val_metrics["loss"]
                self.metrics.best_epoch = epoch + 1
                if self.config.save_best_only:
                    self._save_checkpoint("best")

            # Early stopping
            if self.early_stopping and self.early_stopping(val_metrics["loss"]):
                print(f"Early stopping triggered at epoch {epoch + 1}")
                break

        # Final save
        self._save_checkpoint("final")

        elapsed = time.time() - start_time
        print(f"\nTraining completed in {elapsed:.2f}s")
        print(f"Best validation loss: {self.metrics.best_val_loss:.4f} "
              f"(epoch {self.metrics.best_epoch})")

        return self.metrics

    def _train_epoch(
        self,
        train_loader: DataLoader,
        optimizer: torch.optim.Optimizer,
        scheduler
    ) -> float:
        """Train for one epoch."""
        self.model.train()
        total_loss = 0.0
        num_batches = 0

        for batch_idx, batch in enumerate(train_loader):
            # Move to device
            input_ids = batch["input_ids"].to(self.device)
            labels = batch["labels"].to(self.device)

            # Create dummy hidden states from input_ids
            # In practice, this would come from a transformer encoder
            hidden_states = torch.randn(
                input_ids.size(0), 128, 768,
                device=self.device
            )

            # Forward pass with mixed precision
            if self.scaler:
                with torch.cuda.amp.autocast():
                    logits = self.model(hidden_states)
                    loss = F.cross_entropy(logits, labels)
                    loss = loss / self.config.gradient_accumulation_steps

                self.scaler.scale(loss).backward()

                if (batch_idx + 1) % self.config.gradient_accumulation_steps == 0:
                    self.scaler.unscale_(optimizer)
                    torch.nn.utils.clip_grad_norm_(
                        self.model.parameters(),
                        self.config.max_grad_norm
                    )
                    self.scaler.step(optimizer)
                    self.scaler.update()
                    optimizer.zero_grad()
                    scheduler.step()
            else:
                logits = self.model(hidden_states)
                loss = F.cross_entropy(logits, labels)
                loss = loss / self.config.gradient_accumulation_steps
                loss.backward()

                if (batch_idx + 1) % self.config.gradient_accumulation_steps == 0:
                    torch.nn.utils.clip_grad_norm_(
                        self.model.parameters(),
                        self.config.max_grad_norm
                    )
                    optimizer.step()
                    optimizer.zero_grad()
                    scheduler.step()

            total_loss += loss.item() * self.config.gradient_accumulation_steps
            num_batches += 1

            self.metrics.step += 1

        return total_loss / num_batches

    def _validate(self, val_loader: DataLoader) -> Dict[str, float]:
        """Validate the model."""
        self.model.eval()
        total_loss = 0.0
        correct = 0
        total = 0

        with torch.no_grad():
            for batch in val_loader:
                input_ids = batch["input_ids"].to(self.device)
                labels = batch["labels"].to(self.device)

                # Create dummy hidden states
                hidden_states = torch.randn(
                    input_ids.size(0), 128, 768,
                    device=self.device
                )

                logits = self.model(hidden_states)
                loss = F.cross_entropy(logits, labels)

                total_loss += loss.item()
                predictions = logits.argmax(dim=-1)
                correct += (predictions == labels).sum().item()
                total += labels.size(0)

        return {
            "loss": total_loss / len(val_loader),
            "accuracy": correct / total
        }

    def evaluate(self, test_data: pd.DataFrame) -> Dict[str, float]:
        """
        Evaluate on test data.

        Args:
            test_data: Test DataFrame with 'text' and 'label' columns

        Returns:
            Dictionary with evaluation metrics
        """
        test_dataset = FinancialTextDataset(
            test_data["text"].tolist(),
            test_data["label"].tolist()
        )
        test_loader = DataLoader(
            test_dataset,
            batch_size=self.config.batch_size
        )

        metrics = self._validate(test_loader)

        # Compute additional metrics
        all_predictions = []
        all_labels = []

        self.model.eval()
        with torch.no_grad():
            for batch in test_loader:
                input_ids = batch["input_ids"].to(self.device)
                labels = batch["labels"]

                hidden_states = torch.randn(
                    input_ids.size(0), 128, 768,
                    device=self.device
                )

                logits = self.model(hidden_states)
                predictions = logits.argmax(dim=-1).cpu()

                all_predictions.extend(predictions.tolist())
                all_labels.extend(labels.tolist())

        # Compute F1 score
        from collections import Counter
        pred_counts = Counter(all_predictions)
        label_counts = Counter(all_labels)

        # Macro F1
        f1_scores = []
        for label in set(all_labels):
            tp = sum(1 for p, l in zip(all_predictions, all_labels)
                    if p == l == label)
            precision = tp / (pred_counts.get(label, 1) or 1)
            recall = tp / (label_counts.get(label, 1) or 1)
            f1 = 2 * precision * recall / (precision + recall + 1e-8)
            f1_scores.append(f1)

        metrics["f1"] = np.mean(f1_scores)
        metrics["num_samples"] = len(all_labels)

        return metrics

    def _save_checkpoint(self, name: str):
        """Save model checkpoint."""
        os.makedirs(self.config.save_dir, exist_ok=True)

        checkpoint_path = os.path.join(
            self.config.save_dir,
            f"checkpoint_{name}.pt"
        )

        torch.save({
            "model_state_dict": self.model.state_dict(),
            "config": asdict(self.config),
            "metrics": asdict(self.metrics)
        }, checkpoint_path)

        print(f"Checkpoint saved: {checkpoint_path}")

    def load_checkpoint(self, path: str):
        """Load model from checkpoint."""
        checkpoint = torch.load(path, map_location=self.device)
        self.model.load_state_dict(checkpoint["model_state_dict"])
        print(f"Loaded checkpoint: {path}")

    def save_model(self, path: str):
        """Save model for deployment."""
        os.makedirs(os.path.dirname(path) or ".", exist_ok=True)

        # Save model state
        torch.save(self.model.state_dict(), path)

        # Save config
        config_path = path.replace(".pt", "_config.json")
        with open(config_path, "w") as f:
            json.dump(asdict(self.config), f, indent=2)

        print(f"Model saved: {path}")


if __name__ == "__main__":
    from .data_loader import load_financial_phrasebank

    # Load sample data
    train_data, val_data = load_financial_phrasebank()

    print(f"Training samples: {len(train_data)}")
    print(f"Validation samples: {len(val_data)}")

    # Create trainer
    trainer = FineTuningTrainer(
        model_name="finbert",
        method="lora",
        lora_rank=8,
        learning_rate=2e-4
    )

    # Train
    metrics = trainer.train(train_data, val_data, epochs=2)

    # Evaluate
    eval_metrics = trainer.evaluate(val_data)
    print(f"\nFinal evaluation:")
    print(f"  Accuracy: {eval_metrics['accuracy']:.4f}")
    print(f"  F1 Score: {eval_metrics['f1']:.4f}")

File: rust/src/trainer.rs
//! Training pipeline for fine-tuning models.
//!
//! This module provides:
//! - `TrainingConfig`: Configuration for training
//! - `TrainingMetrics`: Metrics tracked during training
//! - `FineTuningTrainer`: Main training loop with early stopping

use crate::model::FinancialSentimentModel;
use ndarray::{Array1, Array2, Axis};
use serde::{Deserialize, Serialize};
use std::time::Instant;

/// Training configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrainingConfig {
    /// Learning rate
    pub learning_rate: f32,
    /// Batch size
    pub batch_size: usize,
    /// Number of training epochs
    pub num_epochs: usize,
    /// Early stopping patience (epochs)
    pub early_stopping_patience: usize,
    /// Weight decay for regularization
    pub weight_decay: f32,
    /// Number of warmup steps for learning rate schedule
    pub warmup_steps: usize,
    /// Maximum gradient norm for clipping
    pub max_grad_norm: f32,
    /// Whether to use gradient checkpointing
    pub gradient_checkpointing: bool,
    /// Random seed
    pub seed: u64,
    /// Logging interval (steps)
    pub log_interval: usize,
    /// Evaluation interval (steps)
    pub eval_interval: usize,
}

impl Default for TrainingConfig {
    fn default() -> Self {
        Self {
            learning_rate: 1e-4,
            batch_size: 32,
            num_epochs: 10,
            early_stopping_patience: 5,
            weight_decay: 0.01,
            warmup_steps: 100,
            max_grad_norm: 1.0,
            gradient_checkpointing: false,
            seed: 42,
            log_interval: 100,
            eval_interval: 500,
        }
    }
}

impl TrainingConfig {
    /// Create a new training configuration with custom learning rate.
    pub fn with_learning_rate(mut self, lr: f32) -> Self {
        self.learning_rate = lr;
        self
    }

    /// Set batch size.
    pub fn with_batch_size(mut self, batch_size: usize) -> Self {
        self.batch_size = batch_size;
        self
    }

    /// Set number of epochs.
    pub fn with_epochs(mut self, epochs: usize) -> Self {
        self.num_epochs = epochs;
        self
    }

    /// Set early stopping patience.
    pub fn with_patience(mut self, patience: usize) -> Self {
        self.early_stopping_patience = patience;
        self
    }
}

/// Metrics tracked during training.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrainingMetrics {
    /// Training loss history
    pub train_losses: Vec<f32>,
    /// Validation loss history
    pub val_losses: Vec<f32>,
    /// Training accuracy history
    pub train_accuracies: Vec<f32>,
    /// Validation accuracy history
    pub val_accuracies: Vec<f32>,
    /// Best validation loss achieved
    pub best_val_loss: f32,
    /// Epoch at which best validation loss was achieved
    pub best_epoch: usize,
    /// Total training time in seconds
    pub training_time_secs: f64,
    /// Number of early stopping triggers
    pub early_stop_count: usize,
    /// Whether training was stopped early
    pub stopped_early: bool,
}

impl Default for TrainingMetrics {
    fn default() -> Self {
        Self {
            train_losses: Vec::new(),
            val_losses: Vec::new(),
            train_accuracies: Vec::new(),
            val_accuracies: Vec::new(),
            best_val_loss: f32::INFINITY,
            best_epoch: 0,
            training_time_secs: 0.0,
            early_stop_count: 0,
            stopped_early: false,
        }
    }
}

/// Fine-tuning trainer with early stopping.
pub struct FineTuningTrainer {
    /// Model to train
    pub model: FinancialSentimentModel,
    /// Training configuration
    pub config: TrainingConfig,
    /// Training metrics
    pub metrics: TrainingMetrics,
    /// Current learning rate
    current_lr: f32,
    /// Global step counter
    global_step: usize,
}

impl FineTuningTrainer {
    /// Create a new trainer.
    pub fn new(model: FinancialSentimentModel, config: TrainingConfig) -> Self {
        let current_lr = config.learning_rate;
        Self {
            model,
            config,
            metrics: TrainingMetrics::default(),
            current_lr,
            global_step: 0,
        }
    }

    /// Get learning rate with warmup schedule.
    fn get_learning_rate(&self) -> f32 {
        if self.global_step < self.config.warmup_steps {
            // Linear warmup
            self.config.learning_rate * (self.global_step as f32 / self.config.warmup_steps as f32)
        } else {
            self.config.learning_rate
        }
    }

    /// Compute cross-entropy loss.
    fn cross_entropy_loss(logits: &Array1<f32>, target: usize) -> f32 {
        // Softmax
        let max_logit = logits.iter().cloned().fold(f32::NEG_INFINITY, f32::max);
        let exp_logits: Array1<f32> = logits.mapv(|x| (x - max_logit).exp());
        let sum_exp = exp_logits.sum();
        let log_softmax = logits.mapv(|x| x - max_logit - sum_exp.ln());

        // Negative log likelihood
        -log_softmax[target]
    }

    /// Compute accuracy.
    fn compute_accuracy(predictions: &[usize], targets: &[usize]) -> f32 {
        let correct = predictions
            .iter()
            .zip(targets.iter())
            .filter(|(p, t)| p == t)
            .count();
        correct as f32 / predictions.len() as f32
    }

    /// Train for one epoch.
    fn train_epoch(
        &mut self,
        train_x: &Array2<f32>,
        train_y: &[usize],
    ) -> (f32, f32) {
        let n_samples = train_x.nrows();
        let batch_size = self.config.batch_size.min(n_samples);
        let n_batches = (n_samples + batch_size - 1) / batch_size;

        let mut total_loss = 0.0;
        let mut all_predictions = Vec::new();

        // Simple random permutation for shuffling
        let mut indices: Vec<usize> = (0..n_samples).collect();
        // Fisher-Yates shuffle
        for i in (1..n_samples).rev() {
            let j = (self.global_step * 1337 + i) % (i + 1);
            indices.swap(i, j);
        }

        for batch_idx in 0..n_batches {
            let start = batch_idx * batch_size;
            let end = (start + batch_size).min(n_samples);

            let mut batch_loss = 0.0;

            for &idx in &indices[start..end] {
                let input = train_x.row(idx).to_owned();
                let target = train_y[idx];

                // Forward pass
                let logits = self.model.forward(&input);
                let loss = Self::cross_entropy_loss(&logits, target);
                batch_loss += loss;

                // Get prediction
                let pred = logits
                    .iter()
                    .enumerate()
                    .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
                    .map(|(i, _)| i)
                    .unwrap_or(0);
                all_predictions.push(pred);

                self.global_step += 1;
            }

            total_loss += batch_loss;
        }

        let avg_loss = total_loss / n_samples as f32;
        let accuracy = Self::compute_accuracy(&all_predictions, train_y);

        (avg_loss, accuracy)
    }

    /// Evaluate on validation data.
    fn evaluate(&self, val_x: &Array2<f32>, val_y: &[usize]) -> (f32, f32) {
        let n_samples = val_x.nrows();
        let mut total_loss = 0.0;
        let mut predictions = Vec::new();

        for i in 0..n_samples {
            let input = val_x.row(i).to_owned();
            let target = val_y[i];

            let logits = self.model.forward(&input);
            let loss = Self::cross_entropy_loss(&logits, target);
            total_loss += loss;

            let pred = logits
                .iter()
                .enumerate()
                .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
                .map(|(i, _)| i)
                .unwrap_or(0);
            predictions.push(pred);
        }

        let avg_loss = total_loss / n_samples as f32;
        let accuracy = Self::compute_accuracy(&predictions, val_y);

        (avg_loss, accuracy)
    }

    /// Run the full training loop.
    ///
    /// # Arguments
    ///
    /// * `train_x` - Training features of shape (n_train, dim)
    /// * `train_y` - Training labels
    /// * `val_x` - Validation features of shape (n_val, dim)
    /// * `val_y` - Validation labels
    ///
    /// # Returns
    ///
    /// Training metrics after completion.
    pub fn train(
        &mut self,
        train_x: &Array2<f32>,
        train_y: &[usize],
        val_x: &Array2<f32>,
        val_y: &[usize],
    ) -> TrainingMetrics {
        let start_time = Instant::now();
        let mut patience_counter = 0;

        println!("Starting training...");
        println!("  Train samples: {}", train_x.nrows());
        println!("  Val samples: {}", val_x.nrows());
        println!("  Epochs: {}", self.config.num_epochs);
        println!("  Batch size: {}", self.config.batch_size);
        println!();

        for epoch in 0..self.config.num_epochs {
            // Training
            let (train_loss, train_acc) = self.train_epoch(train_x, train_y);
            self.metrics.train_losses.push(train_loss);
            self.metrics.train_accuracies.push(train_acc);

            // Validation
            let (val_loss, val_acc) = self.evaluate(val_x, val_y);
            self.metrics.val_losses.push(val_loss);
            self.metrics.val_accuracies.push(val_acc);

            // Print progress
            println!(
                "Epoch {}/{}: train_loss={:.4}, train_acc={:.4}, val_loss={:.4}, val_acc={:.4}",
                epoch + 1,
                self.config.num_epochs,
                train_loss,
                train_acc,
                val_loss,
                val_acc
            );

            // Check for improvement
            if val_loss < self.metrics.best_val_loss {
                self.metrics.best_val_loss = val_loss;
                self.metrics.best_epoch = epoch;
                patience_counter = 0;
            } else {
                patience_counter += 1;

                if patience_counter >= self.config.early_stopping_patience {
                    println!(
                        "\nEarly stopping triggered at epoch {} (patience={})",
                        epoch + 1,
                        self.config.early_stopping_patience
                    );
                    self.metrics.stopped_early = true;
                    break;
                }
            }
        }

        self.metrics.training_time_secs = start_time.elapsed().as_secs_f64();
        self.metrics.early_stop_count = patience_counter;

        println!("\nTraining completed!");
        println!("  Best val_loss: {:.4} at epoch {}", self.metrics.best_val_loss, self.metrics.best_epoch + 1);
        println!("  Training time: {:.2}s", self.metrics.training_time_secs);

        self.metrics.clone()
    }

    /// Get the trained model.
    pub fn get_model(&self) -> &FinancialSentimentModel {
        &self.model
    }

    /// Take ownership of the trained model.
    pub fn into_model(self) -> FinancialSentimentModel {
        self.model
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::Array2;
    use ndarray_rand::RandomExt;
    use rand_distr::Normal;

    #[test]
    fn test_training_config_default() {
        let config = TrainingConfig::default();
        assert_eq!(config.learning_rate, 1e-4);
        assert_eq!(config.batch_size, 32);
        assert_eq!(config.num_epochs, 10);
    }

    #[test]
    fn test_training_config_builder() {
        let config = TrainingConfig::default()
            .with_learning_rate(1e-3)
            .with_batch_size(64)
            .with_epochs(20);

        assert_eq!(config.learning_rate, 1e-3);
        assert_eq!(config.batch_size, 64);
        assert_eq!(config.num_epochs, 20);
    }

    #[test]
    fn test_trainer_creation() {
        let model = FinancialSentimentModel::new(768, 12, 8, 16.0, 3);
        let config = TrainingConfig::default();
        let trainer = FineTuningTrainer::new(model, config);

        assert_eq!(trainer.global_step, 0);
    }

    #[test]
    fn test_cross_entropy_loss() {
        let logits = Array1::from_vec(vec![2.0, 1.0, 0.1]);
        let loss = FineTuningTrainer::cross_entropy_loss(&logits, 0);
        assert!(loss > 0.0);
        assert!(loss < 2.0);
    }

    #[test]
    fn test_short_training() {
        let model = FinancialSentimentModel::new(64, 4, 4, 8.0, 3);
        let config = TrainingConfig::default()
            .with_epochs(2)
            .with_batch_size(4);

        let mut trainer = FineTuningTrainer::new(model, config);

        // Create small synthetic dataset
        let normal = Normal::new(0.0, 1.0).unwrap();
        let train_x = Array2::random((20, 64), normal.clone());
        let train_y: Vec<usize> = (0..20).map(|i| i % 3).collect();
        let val_x = Array2::random((10, 64), normal);
        let val_y: Vec<usize> = (0..10).map(|i| i % 3).collect();

        let metrics = trainer.train(&train_x, &train_y, &val_x, &val_y);

        assert_eq!(metrics.train_losses.len(), 2);
        assert_eq!(metrics.val_losses.len(), 2);
    }
}
